### PR TITLE
Feature/Ensure Idempotency for UserInput Submission

### LINKED ISSUE #219: Implement Idempotency for UserInput Submission
**Related User Story**
#120 

**Description**
https://github.com/TanStack/query/discussions/608
This article discusses some TanStack Query default behavior that may not be working in Linkta.

Implement idempotency in the `POST /v1/inputs` endpoint to ensure multiple identical requests do not result in duplicate LinktaFlow objects or unintended side effects.

The stated endpoint currently processes UserInput to generate a LinktaFlow. To enhance reliability and prevent duplicate processing idempotency implementation is needed. This will ensure multiple identical requests with the same `x-request-id` header result in the same outcome.

Note: Please research the topic and discuss it with the team for feedback.

**Acceptance Criteria**
- Multiple requests with the same `x-request-id` return the same response and do not create duplicate LinktaFlow objects.
- If time allows, keep the logic encapsulated so it can be reused in another endpoint
- Appropriate error messages are returned for invalid or missing `x-request-id`.

### PR DESCRIPTION
**Pull request by kinskiwu**
## Description
Frontend (Throttling):
- [X] Implement custom useThrottle hook
- [X] Apply useThrottle hook to form submission function in UserInputForm (1s)
- [X] Update button disabled state to reflect throttling status
- [X] Enhance error message on UserInputForm SnackBar

Backend (Idempotency):
- [X] Create data model for IdempotencyRecordSchema (add cleanup mechanism for expired records)
- [X] Create corresponding TypeScript type for IdempotencyTracker
- [X] Implement createIdempotencyService
- [X] Create idempotency middleware
- [X] Insert idempotency middleware into POST /v1/inputs route
- [X] Implement error handling
- [X] Add logging
- [X] Update API documentation

Others
- [X] Increase token limit for Gemini from 1000 to 1500
- [X] Increase rate limit from 60 to 500 per 15mins for development 

Fixes #219 

## Type of change
- [X] ✨  New feature (non-breaking change which adds functionality)
- [X] 📝 Documentation update

## How Can this be tested? Testing Plan to review this PR
- [X] Where in the app does this change take place?
- [X] What are the steps to test this change?
- [X] What is the expected output?
1. **Comment out line 133 in `/client/components/user-input/UserInputForm.tsx`.**
2. **Spin up the app at the root directory.**
3. **Open Chrome DevTools in your browser.**
4. **Test the throttling functionality:**
   - Click on `Explore a New Topic`.
   - Insert a prompt.
   - Click `Generate` multiple times within 1 second.
   - (To increase the delay time, update line 78 in `UserInputForm.tsx`.)
5. **Observe the Console Output:**
   - You should see only one response from the API.
   - If an error occurs, a detailed error message will be displayed in the SnackBar notification at the bottom left of the page.
6. **How I tested the idempotency functionality using thunder client:**
   - Comment out line 54-61 in `server.ts`
   - Replace line 37 in `userInputController.ts` with `MOCK_USER_ID` from `@mocks/userId`
   - Open thunder client -> Body -> JSON:
   ```json
   {
     "input": "web development"
   }
   ```
   - Headers -> `x-request-id`: <your choice of new id`
   - Initial request with `x-request-id` 123:
Pasted Graphic 3
   - Subsequent requests with the same id:
Pasted Graphic 4

## Checklist:

- [X] I have performed a self-review of my own code.
- [X] I have commented on my code, particularly in hard-to-understand areas.
- [X] I have made corresponding changes to the documentation.
- [X] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [X] I have reviewed the [**Security Best Practices Document**](./docs/SECURITY_BEST_PRACTICES.md).
- [X] I have followed the [**Naming Conventions Guide**](./docs/NAMING_CONVENTIONS_GUIDE.md) for my changes.
- [X] I have ensured that my pull request title is descriptive.

### CODE REVIEW

**Review by colinvandergraaf — APPROVED** [review]
There's a lot of great stuff in this PR, and it looks like a really comprehensive solution.  This is the first time i've seen it implemented in a project. I'm curious about the pattern of implementing both throttling the requests client side and use the idempotency key to handle duplicated requests on the backend. Wouldn't the backend implementation be enough? Is the useThrottle really necessary here? 

Also thanks for having such detailed review instructions on this 🙌

**Comment by kinskiwu** [comment]
Thank you @colinvandergraaf. Thats a great question.

The purpose of the useThrottle hook is to make sure only 1 api call is made within x amount of time. It helps reduce server load and network traffic in the case of rapid fire requests or accidental duplicated clicks.

The idempotency middleware on the other hand ensures the same request marked with the x-request-id only creates 1 document in the db. This helps prevent accidental document creation caused by network issues or other unintended causes.

While one technique helps in some scenarios, having both gives us a better protection. Hope it helps answer your question.

File: client/components/user-input/UserInputForm.tsx [hunk 1]
```diff
@@ -121,10 +130,10 @@ const UserInputForm = () => {
           variant='contained'
           className={`${styles.userInputSubmitButton}`}
           type='submit'
-          disabled={!isChecked}
+          disabled={!isChecked || createLinktaFlowMutation.status === 'pending'}
```
Comment (tedgusek) [inline client/components/user-input/UserInputForm.tsx:133]: Nice catch here

**Review by tedgusek — COMMENTED** [review]
This is my first time looking over this, but it appears to be clearly written, and looks good to me!

**Review by tedgusek — APPROVED** [review]
Sorry, I left the previous comment as a comment, and not an approval.

**Pull request merged**